File: files/000-099/023/CH-023.R
library(tidyverse)
library(readxl)

input1 = read_excel("files/CH-023 Advance Weighted AVG.xlsx", range = "B2:E18")
input2 = read_excel("files/CH-023 Advance Weighted AVG.xlsx", range = "G2:J5")
test   = read_excel("files/CH-023 Advance Weighted AVG.xlsx", range = "L2:M5")

prod = input2 %>%
  pivot_longer(cols = -Month, names_to = "Machine Code", values_to = "value") 

result = input1 %>%
  group_by(Month, `Machine Code`) %>%
  summarise(Avg = mean(`Weight (KG/Meter)`)) %>%
  left_join(prod, by = c("Machine Code", "Month")) %>%
  ungroup() %>%
  group_by(Month) %>%
  summarise(`AVG weight (Kg/Meter)` = sum(Avg * value) / sum(value)) %>%
  mutate(`AVG weight (Kg/Meter)` = round(`AVG weight (Kg/Meter)`, 2)) %>%
  ungroup()

identical(result, test)
# [1] TRUE

File: files/000-099/023/CH-023.py
import pandas as pd

input1 = pd.read_excel("CH-023 Advance Weighted AVG.xlsx", usecols="B:E", skiprows=1, nrows=17)
input2 = pd.read_excel("CH-023 Advance Weighted AVG.xlsx", usecols="G:J", skiprows=1, nrows=4)
test = pd.read_excel("CH-023 Advance Weighted AVG.xlsx", usecols="L:M", skiprows=1, nrows=4)

prod = input2.melt(id_vars="Month", var_name="Machine Code", value_name="value")
result = (
    input1.groupby(["Month", "Machine Code"], as_index=False)["Weight (KG/Meter)"]
    .mean()
    .rename(columns={"Weight (KG/Meter)": "Avg"})
    .merge(prod, on=["Machine Code", "Month"], how="left")
    .groupby("Month", as_index=False)
    .apply(lambda g: pd.Series({"AVG weight (Kg/Meter)": round((g["Avg"] * g["value"]).sum() / g["value"].sum(), 2)}))
    .reset_index(drop=True)
)

print(result.equals(test))
